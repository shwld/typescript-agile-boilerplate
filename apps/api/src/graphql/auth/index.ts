import {
  GenericAuthPluginOptions,
  ResolveUserFn,
  useGenericAuth,
  ValidateUserFn,
} from '@envelop/generic-auth';
import { GraphqlServerContext } from '../context';

type UserType = {
  id: string;
};

const resolveUserFn: ResolveUserFn<UserType, GraphqlServerContext> = async (
  context
) => {
  // Here you can implement any custom sync/async code, and use the context built so far in Envelop and the HTTP request
  // to find the current user.
  // Common practice is to use a JWT token here, validate it, and use the payload as-is, or fetch the user from an external services.
  // Make sure to either return `null` or the user object.

  try {
    // const user = await context.authApi.authenticateUser(
    //   context.req.headers.authorization
    // );

    // return user;
    return { id: 'test' };
  } catch (e) {
    console.error('Failed to validate token');

    return null;
  }
};

const validateUser: ValidateUserFn<UserType> = ({ user }) => {
  // Here you can implement any custom to check if the user is valid and have access to the server.
  // This method is being triggered in different flows, based on the mode you chose to implement.

  // If you are using the `protect-auth-directive` mode, you'll also get 2 additional parameters: the resolver parameters as object and the DirectiveNode of the auth directive.

  if (!user) {
    throw new Error(`Unauthenticated!`);
  }
};

export const useAuth = (
  options?: Omit<
    GenericAuthPluginOptions<UserType, GraphqlServerContext>,
    'resolveUserFn' | 'validateUser'
  >
) =>
  useGenericAuth({
    resolveUserFn,
    validateUser,
    mode: 'protect-all',
    ...options,
  });
