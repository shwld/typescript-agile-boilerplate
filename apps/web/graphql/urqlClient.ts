import { withUrqlClient, WithUrqlClientOptions } from 'next-urql';
import { dedupExchange, cacheExchange, fetchExchange } from '@urql/core';

export const withClient = (options: WithUrqlClientOptions = { ssr: true }) =>
  withUrqlClient(
    () => ({
      url: 'http://localhost:3000/api/graphql',
      fetchOptions: {
        headers: {
          'Content-Type': 'application/json',
        },
      },
      exchanges: [dedupExchange, cacheExchange, fetchExchange],
    }),
    options
  );
