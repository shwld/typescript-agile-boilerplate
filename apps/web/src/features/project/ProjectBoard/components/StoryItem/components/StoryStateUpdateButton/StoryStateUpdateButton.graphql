fragment StoryStateUpdateButton_Story on Story {
  id
  state
}

mutation StoryStateUpdateButton_UpdateStoryState(
  $input: UpdateStoryStateInput!
) {
  updateStoryState(input: $input) {
    ... on UpdateStoryStateSuccessResult {
      result {
        ...StoryStateUpdateButton_Story
      }
    }
    ... on InvalidArgumentsResult {
      issues {
        field
        message
      }
    }
  }
}
