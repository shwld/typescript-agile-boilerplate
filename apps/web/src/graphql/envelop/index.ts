/* eslint-disable react-hooks/rules-of-hooks */
import { envelop, useLogger, useSchema, useTiming } from '@envelop/core';
import { useGraphQLMiddleware } from '@envelop/graphql-middleware';
import { permissions, schema } from 'graphql-resolvers';
import { useOrderedServerContextPlugins } from './context';

export const getEnveloped = envelop({
  plugins: [
    useSchema(schema),
    ...useOrderedServerContextPlugins(),
    useLogger(),
    useTiming(),
    useGraphQLMiddleware([permissions]),
  ],
});
