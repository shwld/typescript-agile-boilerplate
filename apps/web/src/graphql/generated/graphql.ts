import gql from 'graphql-tag';
import * as Urql from 'urql';
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  DateTime: any;
};

export type Connection = {
  edges?: Maybe<Array<Maybe<Edge>>>;
  pageInfo?: Maybe<PageInfo>;
};

export type CreateTodoInput = {
  id: Scalars['ID'];
  title: Scalars['String'];
};

export type Edge = {
  cursor?: Maybe<Scalars['String']>;
  node?: Maybe<Node>;
};

export type InvalidArgumentsResult = {
  __typename?: 'InvalidArgumentsResult';
  issues: Array<ValidationIssue>;
};

export type Mutation = {
  __typename?: 'Mutation';
  createTodo: TodoMutationResult;
};


export type MutationCreateTodoArgs = {
  input: CreateTodoInput;
};

export type Node = {
  id: Scalars['ID'];
};

export type PageInfo = {
  __typename?: 'PageInfo';
  endCursor?: Maybe<Scalars['String']>;
  hasNextPage: Scalars['Boolean'];
  hasPreviousPage?: Maybe<Scalars['Boolean']>;
  startCursor?: Maybe<Scalars['String']>;
};

export type PagedConnection = {
  nodes?: Maybe<Array<Maybe<Node>>>;
  pageInfo?: Maybe<PagedPageInfo>;
};

export type PagedPageInfo = {
  __typename?: 'PagedPageInfo';
  hasNextPage: Scalars['Boolean'];
  hasPreviousPage?: Maybe<Scalars['Boolean']>;
  totalPagesCount?: Maybe<Scalars['Int']>;
};

export type Query = {
  __typename?: 'Query';
  node?: Maybe<Node>;
  viewer?: Maybe<Viewer>;
};


export type QueryNodeArgs = {
  id: Scalars['ID'];
};

export type Todo = Node & {
  __typename?: 'Todo';
  createdAt: Scalars['DateTime'];
  id: Scalars['ID'];
  title: Scalars['String'];
  updatedAt: Scalars['DateTime'];
};

export type TodoConnection = Connection & {
  __typename?: 'TodoConnection';
  edges?: Maybe<Array<Maybe<TodoEdge>>>;
  pageInfo?: Maybe<PageInfo>;
};

export type TodoEdge = Edge & {
  __typename?: 'TodoEdge';
  cursor?: Maybe<Scalars['String']>;
  node?: Maybe<Todo>;
};

export type TodoMutationResult = InvalidArgumentsResult | TodoSuccessResult | UnauthenticatedResult;

export type TodoSuccessResult = {
  __typename?: 'TodoSuccessResult';
  result: Todo;
};

export type UnauthenticatedResult = {
  __typename?: 'UnauthenticatedResult';
  errorMessage: Scalars['String'];
};

export type User = {
  __typename?: 'User';
  id: Scalars['ID'];
  name: Scalars['String'];
};

export type ValidationIssue = {
  __typename?: 'ValidationIssue';
  field?: Maybe<Scalars['String']>;
  message?: Maybe<Scalars['String']>;
};

export type Viewer = {
  __typename?: 'Viewer';
  createdAt: Scalars['DateTime'];
  email: Scalars['String'];
  id: Scalars['ID'];
  picture: Scalars['String'];
  todos: TodoConnection;
  updatedAt: Scalars['DateTime'];
};


export type ViewerTodosArgs = {
  after?: InputMaybe<Scalars['String']>;
  first?: InputMaybe<Scalars['Int']>;
};

export type TodoCreateFormResultFragment = { __typename?: 'Todo', id: string, title: string };

export type TodoCreateFormMutationVariables = Exact<{
  input: CreateTodoInput;
}>;


export type TodoCreateFormMutation = { __typename?: 'Mutation', createTodo: { __typename?: 'InvalidArgumentsResult' } | { __typename?: 'TodoSuccessResult', result: { __typename?: 'Todo', id: string, title: string } } | { __typename?: 'UnauthenticatedResult' } };

export type TodoListResultFragment = { __typename?: 'Todo', id: string, title: string };

export type TodoListQueryVariables = Exact<{
  cursor?: InputMaybe<Scalars['String']>;
}>;


export type TodoListQuery = { __typename?: 'Query', viewer?: { __typename?: 'Viewer', id: string, todos: { __typename?: 'TodoConnection', edges?: Array<{ __typename?: 'TodoEdge', cursor?: string | null, node?: { __typename?: 'Todo', id: string, title: string } | null } | null> | null, pageInfo?: { __typename?: 'PageInfo', hasNextPage: boolean, endCursor?: string | null } | null } } | null };

export type SampleQueryVariables = Exact<{ [key: string]: never; }>;


export type SampleQuery = { __typename?: 'Query', viewer?: { __typename?: 'Viewer', id: string, email: string, picture: string } | null };

export const TodoCreateFormResult = gql`
    fragment TodoCreateFormResult on Todo {
  id
  title
}
    `;
export const TodoListResult = gql`
    fragment TodoListResult on Todo {
  id
  title
}
    `;
export const TodoCreateForm = gql`
    mutation todoCreateForm($input: CreateTodoInput!) {
  createTodo(input: $input) {
    ... on TodoSuccessResult {
      result {
        ...TodoCreateFormResult
      }
    }
  }
}
    ${TodoCreateFormResult}`;
export const TodoList = gql`
    query todoList($cursor: String) {
  viewer {
    id
    todos(first: 3, after: $cursor) {
      edges {
        node {
          ...TodoListResult
        }
        cursor
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
}
    ${TodoListResult}`;
export const Sample = gql`
    query sample {
  viewer {
    id
    email
    picture
  }
}
    `;
export const TodoCreateFormResultFragmentDoc = gql`
    fragment TodoCreateFormResult on Todo {
  id
  title
}
    `;
export const TodoListResultFragmentDoc = gql`
    fragment TodoListResult on Todo {
  id
  title
}
    `;
export const TodoCreateFormDocument = gql`
    mutation todoCreateForm($input: CreateTodoInput!) {
  createTodo(input: $input) {
    ... on TodoSuccessResult {
      result {
        ...TodoCreateFormResult
      }
    }
  }
}
    ${TodoCreateFormResultFragmentDoc}`;

export function useTodoCreateFormMutation() {
  return Urql.useMutation<TodoCreateFormMutation, TodoCreateFormMutationVariables>(TodoCreateFormDocument);
};
export const TodoListDocument = gql`
    query todoList($cursor: String) {
  viewer {
    id
    todos(first: 3, after: $cursor) {
      edges {
        node {
          ...TodoListResult
        }
        cursor
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
}
    ${TodoListResultFragmentDoc}`;

export function useTodoListQuery(options?: Omit<Urql.UseQueryArgs<TodoListQueryVariables>, 'query'>) {
  return Urql.useQuery<TodoListQuery>({ query: TodoListDocument, ...options });
};
export const SampleDocument = gql`
    query sample {
  viewer {
    id
    email
    picture
  }
}
    `;

export function useSampleQuery(options?: Omit<Urql.UseQueryArgs<SampleQueryVariables>, 'query'>) {
  return Urql.useQuery<SampleQuery>({ query: SampleDocument, ...options });
};