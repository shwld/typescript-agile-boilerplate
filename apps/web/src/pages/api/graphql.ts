import {
  getGraphQLParameters,
  processRequest,
  renderGraphiQL,
  sendResult,
  shouldRenderGraphiQL,
} from 'graphql-helix';
import { NextApiHandler } from 'next/types';
import { prepareUser, schema } from 'graphql-resolvers';
import type { GraphqlServerContext } from 'graphql-resolvers';
import { getToken } from 'next-auth/jwt';
import { db } from 'db';

export default (async (req, res) => {
  const token = await getToken({
    req,
    secret: process.env.NEXTAUTH_SECRET,
  });
  const user =
    token?.sub != null
      ? {
          id: token.sub,
          email: token.email ?? '',
          name: token.name ?? 'No name',
          picture: token.picture ?? '',
        }
      : undefined;

  const context: GraphqlServerContext = {
    user: user != null ? await prepareUser(db, user) : undefined,
    db,
  };

  const request = {
    body: req.body,
    headers: req.headers,
    method: req.method ?? 'GET',
    query: req.query,
  };

  if (shouldRenderGraphiQL(request)) {
    res.send(renderGraphiQL({ endpoint: '/api/graphql' }));
  } else {
    const { operationName, query, variables } = getGraphQLParameters(request);

    const result = await processRequest<GraphqlServerContext>({
      operationName,
      query,
      variables,
      request,
      schema,
      contextFactory: () => context,
    });

    sendResult(result, res);
  }
}) as NextApiHandler;
