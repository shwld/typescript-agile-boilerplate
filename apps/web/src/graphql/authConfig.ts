import { AuthConfig, authExchange } from '@urql/exchange-auth';
import { IdTokenResult } from 'firebase/auth';
import { makeOperation } from 'urql';

type AuthState = {
  token: string;
  result: IdTokenResult;
};

const authConfig: AuthConfig<AuthState> = {
  async getAuth() {
    // if (authApp.currentUser != null) {
    //   const tokenResult = await authApp.currentUser.getIdTokenResult();
    //   return { token: tokenResult.token, result: tokenResult };
    // }
    return null;
  },
  willAuthError: ({ authState }) => {
    if (authState?.result?.expirationTime) {
      const expirationDate = new Date(authState.result.expirationTime);
      console.log('⚠️ expirationDate: ', expirationDate);
      return expirationDate < new Date();
    }

    return !authState || !authState.token;
  },
  addAuthToOperation: ({ authState, operation }) => {
    if (!authState || !authState.token) {
      return operation;
    }

    const fetchOptions =
      typeof operation.context.fetchOptions === 'function'
        ? operation.context.fetchOptions()
        : operation.context.fetchOptions || {};

    return makeOperation(operation.kind, operation, {
      ...operation.context,
      fetchOptions: {
        ...fetchOptions,
        headers: {
          ...fetchOptions.headers,
          Authorization: 'Bearer ' + authState.token,
        },
      },
    });
  },
};

export const firebaseAuthExchange = authExchange<AuthState>(authConfig);
