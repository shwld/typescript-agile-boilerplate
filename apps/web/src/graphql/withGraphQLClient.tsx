import { withUrqlClient, WithUrqlClientOptions } from 'next-urql';
import { cacheExchange } from '@urql/exchange-graphcache';
import { relayPagination } from '@urql/exchange-graphcache/extras';
import { dedupExchange, fetchExchange } from '@urql/core';
import { NextPage } from 'next';
import App from 'next/app';
import type {
  AccountCreateButtonMutation,
  AccountListQuery,
  AccountListQueryVariables,
  ProjectCreateButtonMutation,
} from './generated/graphql';
import { AccountListDocument } from '~/features/account/AccountList/AccountList.generated';

const cache = cacheExchange({
  resolvers: {
    Viewer: {
      accounts: relayPagination(),
    },
  },
  updates: {
    Mutation: {
      createAccount(result: AccountCreateButtonMutation, _args, cache, _info) {
        if (result.createAccount.__typename !== 'CreateAccountSuccessResult')
          return;
        const node = result.createAccount.result;
        cache.updateQuery<AccountListQuery, AccountListQueryVariables>(
          { query: AccountListDocument, variables: { cursor: '' } },
          (data) => {
            data?.viewer?.accounts.edges?.unshift({
              node,
              cursor: '',
              __typename: 'AccountEdge',
            });
            return data;
          }
        );
      },
      createProject(result: ProjectCreateButtonMutation, _args, cache, _info) {
        if (result.createProject.__typename !== 'CreateProjectSuccessResult')
          return;
        const node = result.createProject.result;
        cache.updateQuery<AccountListQuery, AccountListQueryVariables>(
          { query: AccountListDocument, variables: { cursor: '' } },
          (data) => {
            data?.viewer?.accounts.edges?.forEach((edge) => {
              if (edge?.node != null && edge.node.id === node.accountId) {
                edge.node.projects.edges?.unshift({
                  node,
                  cursor: '',
                  __typename: 'ProjectEdge',
                });
              }
            });
            return data;
          }
        );
      },
    },
  },
});

export const withGraphQLClient = <C extends NextPage<any, any> | typeof App>(
  c: C,
  options: WithUrqlClientOptions = { ssr: false }
) =>
  withUrqlClient((_ssrExchange, _ctx) => {
    return {
      url: `${
        typeof window === 'undefined' ? '' : 'http://localhost:3000'
      }/api/graphql`,
      suspense: true,
      fetchOptions: {
        headers: {
          'Content-Type': 'application/json',
        },
      },
      exchanges: [dedupExchange, cache, fetchExchange],
    };
  }, options)(c);
