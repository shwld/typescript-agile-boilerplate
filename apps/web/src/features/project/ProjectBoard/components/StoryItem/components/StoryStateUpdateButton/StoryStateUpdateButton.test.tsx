import { StoryStateUpdateButton } from './StoryStateUpdateButton';
import { render } from '@testing-library/react';
import { MockedUrqlProvider } from '~/test/MockedUrqlProvider';
import { StoryState } from '~/graphql/generated/graphql';

describe('StoryStateUpdateButton', () => {
  const renderComponent = () => {
    const renderResult = render(
      <MockedUrqlProvider>
        <StoryStateUpdateButton storyId="test" state={StoryState.Started} />
      </MockedUrqlProvider>
    );
    return renderResult;
  };
  test('Snapshot', () => {
    expect(renderComponent().asFragment()).toMatchSnapshot();
  });
  test('success', () => {
    const { getByText } = renderComponent();
    expect(getByText('Finish')).toBeTruthy();
  });
});
