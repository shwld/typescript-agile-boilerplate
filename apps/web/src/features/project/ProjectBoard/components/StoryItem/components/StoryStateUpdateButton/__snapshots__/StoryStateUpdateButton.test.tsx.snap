// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryStateUpdateButton Snapshot 1`] = `
<DocumentFragment>
  <div
    class="chakra-button__group css-1ind3v2"
    role="group"
  >
    <button
      class="chakra-button css-2qv00o"
      type="button"
    >
      Finish
    </button>
  </div>
</DocumentFragment>
`;
